drivers/ethernet/eth_w5500.c

```diff
@@ -0,0 +1,586 @@
+	};
+	const struct spi_buf tx_buf[2] = {
+		{
+			.buf = cmd,
+			.len = ARRAY_SIZE(cmd),
+		},
+		{
+			.buf = data,
+			.len = len,
+		},
+	};
+	const struct spi_buf_set tx = {
+		.buffers = tx_buf,
+		.count = ARRAY_SIZE(tx_buf),
+	};
+
+	k_sem_take(&ctx->tx_rx_sem, K_FOREVER);
+	ret = spi_write(ctx->spi, &ctx->spi_cfg, &tx);
+	k_sem_give(&ctx->tx_rx_sem);
+
+	return ret;
+}
+
+static int w5500_readbuf(struct device *dev, uint16_t offset, uint8_t *buf,
+			 int len)
+{
+	uint32_t addr;
+	int remain = 0;
+	int ret;
+	const uint32_t mem_start = W5500_Sn_RX_MEM_START;
+	const uint16_t mem_size = W5500_RX_MEM_SIZE;
+
+	offset %= mem_size;
+	addr = mem_start + offset;
+
+	if (offset + len > mem_size) {
+		remain = (offset + len) % mem_size;
+		len = mem_size - offset;
+	}
+
+	ret = w5500_spi_read(dev, addr, buf, len);
+	if (ret || !remain) {
+		return ret;
+	}
+
+	return w5500_spi_read(dev, mem_start, buf + len, remain);
+}
+
+static int w5500_writebuf(struct device *dev, uint16_t offset, uint8_t *buf,
+			  int len)
+{
+	uint32_t addr;
+	int ret = 0;
+	int remain = 0;
+	const uint32_t mem_start = W5500_Sn_TX_MEM_START;
+	const uint32_t mem_size = W5500_TX_MEM_SIZE;
+
+	offset %= mem_size;
+	addr = mem_start + offset;
+
+	if (offset + len > mem_size) {
+		remain = (offset + len) % mem_size;
+		len = mem_size - offset;
+	}
+
+	ret = w5500_spi_write(dev, addr, buf, len);
+	if (ret || !remain) {
+		return ret;
+	}
+
+	return w5500_spi_write(dev, mem_start, buf + len, remain);
+}
+
+static int w5500_command(struct device *dev, uint8_t cmd)
+{
+	uint8_t reg;
+	uint64_t end = z_timeout_end_calc(K_MSEC(100));
+
+	w5500_spi_write(dev, W5500_S0_CR, &cmd, 1);
+	do {
+		w5500_spi_read(dev, W5500_S0_CR, &reg, 1);
+		if (end - z_tick_get() <= 0) {
+			return -EIO;
+		}
+
+		k_msleep(1);
+	} while (reg != 0);
+
+	return 0;
+}
+
+static int w5500_tx(struct device *dev, struct net_pkt *pkt)
+{
+	struct w5500_runtime *ctx = dev->data;
+	uint16_t len = net_pkt_get_len(pkt);
+	uint16_t offset;
+	uint8_t off[2];
+	int ret;
+
+	w5500_spi_read(dev, W5500_S0_TX_WR, off, 2);
+	offset = sys_get_be16(off);
+
+	if (net_pkt_read(pkt, ctx->buf, len)) {
+		return -EIO;
+	}
+
+	ret = w5500_writebuf(dev, offset, ctx->buf, len);
+	if (ret < 0) {
+		return ret;
+	}
+
+	sys_put_be16(offset + len, off);
+	ret = w5500_spi_write(dev, W5500_S0_TX_WR, off, 2);
+	w5500_command(dev, S0_CR_SEND);
+
+	return 0;
+}
+
+static void w5500_isr(struct device *dev)
+{
+	uint8_t ir;
+	uint8_t mask = 0;
+	uint8_t header[2];
+	uint8_t tmp[2];
+	uint16_t rx_buf_len;
+	uint16_t rx_len;
+	uint16_t off;
+	struct w5500_runtime *ctx = dev->data;
+	const struct w5500_config *config = dev->config;
+	struct net_pkt *pkt;
+
+	while (true) {
+		k_sem_take(&ctx->int_sem, K_FOREVER);
+
+		w5500_spi_read(dev, W5500_S0_IR, &ir, 1);
+		if (!ir) {
+			goto done;
+		}
+
+		w5500_spi_write(dev, W5500_S0_IR, &ir, 1);
+
+		if (ir & S0_IR_SENDOK) {
+			LOG_DBG("TX Done");
+		}
+
+		if (ir & S0_IR_RECV) {
+			int read_len;
```

Comment:
you could move all the logic in this if () into w5500_rx(), it would be more readable (less indentation, etc...)

Reply:
Moved into w5500_rx now